Integrate Lightbeam

Vurich opened the pull request:
This PR uses Lightbeam (my current working branch of it https://github.com/Vurich/lightbeam/tree/loops) as the compilation backend for wasmtime. Currently this causes most of the tests to fail, since I haven't implemented any floating-point instructions, `br_table`, `select`, globals, or any of the smaller load instructions (such as `i32.load8`), but some of the tests are succeeding and traps appear to work, which was the main thing I was hoping for since I definitely didn't want to duplicate the trap work done here in Lightbeam proper. There's currently an intermittent memory access violation that causes the program to crash when run without `--test-threads 1`, I don't have a good way of finding out what is causing this because memcheck gives several thousand errors, presumably because of deliberate faults being triggered by the compiled WebAssembly.

Also, since Lightbeam implements signature checking in an incompatible way I just have it panic when trying to access the function table.

To test, run `cargo test --features lightbeam -- --test-threads 1`

Vurich commented:
Any idea why the build fails? It looks like it's not checking the `lightbeam` submodule out but it's working when I try to check the repo out in a new folder

lqd commented:
Would using git dependencies in Cargo.toml instead of git submodules make it easier / make it work ?

Vurich commented:
Yes, it would, but I thought it would be easier to keep the two repositories in sync if I included it as a submodule

maxmcd commented on lightbeam:
```diff
@@ -0,0 +1 @@
+/home/jef/Documents/GitHub/Vurich/lightbeam
```
It looks like your submodule is pointing to a local directory. Should likely be pointed at the git repo.

Vurich commented:
Fixed

sunfishcode commented on lib/environ/src/compilation.rs:
```diff
@@ -1,24 +1,77 @@
+            buffer.extend(f);
+            let end = buffer.len();
+            out_map.push(start..end);
+        }
```
Ok, so this is taking the functions that Cranelift compiled into separate `Vec`s, and concatenating them into one big `Vec`, and remembering the indices. And this is because that's what Lightbeam produces. I'll need to think about this a bit more :-).

Vurich replied:
Fixed, we now convert Lightbeam's format into Cranelift's format instead of the other way around.

Vurich commented:
Now that Lightbeam is passing the spec tests, I've updated this branch for the latest master, although debuginfo won't work since we just ignore the flag. I've also cleaned it up based on your feedback. I consider this ready to merge.

Vurich commented on lib/environ/src/lightbeam.rs:
```diff
@@ -0,0 +1,134 @@
+use lightbeam::{self, ModuleContext, Signature};
+
+#[allow(dead_code)]
+mod disassemble {
```
Fixed by removing the module entirely.

Vurich commented on lib/environ/Cargo.toml:
```diff
@@ -15,6 +15,8 @@ edition = "2018"
 cranelift-codegen = "0.28.0"
 cranelift-entity = "0.28.0"
 cranelift-wasm = "0.28.0"
+capstone = "*"
```
Fixed by removing the dependency.

sunfishcode commented:
Sounds good. Let's merge this now, and development can proceed from there! The CI failure is fixed on master already, so we can ignore that here.
